src/models/hf/checkpoint.rs

```diff
@@ -0,0 +1,149 @@
+    Download { source: BoxedError, name: String },
+
+    #[snafu(display("Cannot open or load checkpoint"))]
+    LoadCheckpoint { source: candle_core::Error },
+
+    #[snafu(display("Checkpoint does not exist: {}", name))]
+    NonExistentCheckpoint { name: String },
+
+    #[snafu(display("Shard does not exist: {}", name))]
+    NonExistentShard { name: String },
+
+    #[snafu(display("Cannot open SafeTensors index file: {}", path.to_string_lossy()))]
+    OpenSafeTensorsIndex { source: io::Error, path: PathBuf },
+
+    #[snafu(display("Cannot parse SafeTensors index file: {}", path.to_string_lossy()))]
+    ParseSafeTensorsIndex {
+        source: serde_json::Error,
+        path: PathBuf,
+    },
+}
+
+impl<R> LoadHFCheckpoint for R
+where
+    R: Repo,
+{
+    fn load_hf_checkpoint(&self) -> Result<Box<dyn SimpleBackend>, BoxedError> {
+        self.load_safetensors()
+    }
+}
+
+/// Private trait for loading safetensors checkpoint.
+///
+/// This trait is used to load a safetensors checkpoint from the repository.
+/// We just use it so that we can implement these methods on `Repo` rather
+/// than having them as standalone functions.
+trait LoadHFSafeTensors {
+    fn load_safetensors(&self) -> Result<Box<dyn SimpleBackend>, BoxedError>;
+
+    fn load_safetensors_multi(&self, index_path: &Path) -> Result<Vec<PathBuf>, BoxedError>;
+
+    fn load_safetensors_single(&self) -> Result<Vec<PathBuf>, BoxedError>;
+}
+
+impl<R> LoadHFSafeTensors for R
+where
+    R: Repo,
+{
+    /// Load a safetensors checkpoint.
+    ///
+    /// This method will first probe if there is a shard index. If there is,
+    /// a sharded checkpoint will be loaded. Otherwise, a single-file
+    /// checkpoint is loaded.
+    fn load_safetensors(&self) -> Result<Box<dyn SimpleBackend>, BoxedError> {
+        let file = self.file(SAFETENSORS_INDEX).context(DownloadSnafu {
+            name: SAFETENSORS_INDEX,
+        })?;
+
+        let paths = match file {
+            // We have a safetensor index, so load from shards.
+            Some(index_path) => self.load_safetensors_multi(&index_path),
+
+            // No index file, so assume that there is a single checkpoint.
+            None => self.load_safetensors_single(),
+        }?;
+
+        Ok(Box::new(unsafe {
+            MmapedSafetensors::multi(&paths).context(LoadCheckpointSnafu)?
+        }))
+    }
+
+    /// Load sharded safetensors checkpoint.
+    fn load_safetensors_multi(&self, index_path: &Path) -> Result<Vec<PathBuf>, BoxedError> {
+        // Parse the shard index.
+        let index_file = BufReader::new(
+            File::open(index_path).context(OpenSafeTensorsIndexSnafu { path: index_path })?,
+        );
+        let index: SafeTensorsIndex = serde_json::from_reader(index_file)
+            .context(ParseSafeTensorsIndexSnafu { path: index_path })?;
+
+        let mut shards = Vec::with_capacity(index.shards().len());
+        for shard_name in index.shards() {
```

Comment:
The current impl of `shards` returns an owned collection, so we're discarding after getting the length.

Reply:
Updated to only get once.